tardis/tardis_portal/api_v2/serializers/experiment.py

```diff
@@ -0,0 +1,178 @@
+"""experiment.py
+
+Serializer for experiment and related fields.
+
+This module contains the Django Rest Framework serializers for the Experiment model
+and related models.
+
+Contributors:
+    Chris Seal <c.seal@auckland.ac.nz>
+"""
+
+from django.conf import settings
+
+from rest_framework import serializers
+
+from tardis.apps.dataclassification.models import ExperimentDataClassification
+from tardis.apps.identifiers.models import ExperimentID
+from tardis.apps.projects.models import Project
+from tardis.tardis_portal.api_v2.serializers.project import ProjectIDSerializer
+from tardis.tardis_portal.api_v2.serializers.schema import (
+    ParameterNameSerializer,
+    SchemaSerializer,
+)
+from tardis.tardis_portal.api_v2.serializers.user import UserSerializer
+from tardis.tardis_portal.auth.decorators import (
+    get_accessible_projects_for_experiment,
+    has_sensitive_access,
+)
+from tardis.tardis_portal.models.experiment import Experiment
+from tardis.tardis_portal.models.parameters import (
+    ExperimentParameter,
+    ExperimentParameterSet,
+)
+
+
+class ExperimentParameterSerializer(serializers.ModelSerializer):
+    name = ParameterNameSerializer()
+
+    class Meta:
+        model = ExperimentParameter
+        fields = [
+            "name",
+            "string_value",
+            "numerical_value",
+            "datetime_value",
+        ]
+
+
+class ExperimentParameterSetSerializer(serializers.ModelSerializer):
+    parameters = serializers.SerializerMethodField("get_safe_parameters")
+    schema = SchemaSerializer()
+
+    class Meta:
+        model = ExperimentParameterSet
+        fields = ["schema", "parameters"]
+
+    def get_safe_parameters(self, parameterset_obj):
+        experiment = parameterset_obj.experiment
+        queryset = ExperimentParameter.objects.filter(parameterset=parameterset_obj)
+        parameters = ExperimentParameterSerializer(
+            queryset, many=True, context=self.context
+        ).data
+        if has_sensitive_access(self.context["request"], experiment.pk, "experiment"):
+            return parameters
+        return [item for item in parameters if item.name.sensitive is not True]
```

Comment:
It may be better to omit sensitive metadata regardless of access? If included, sensitive metadata would be included in the starter files intended for shared lab storage, which is viewable to lab members that may not have authorisation to see them. This is unlike including sensitive metadata in RO-Crate which allows for encryption.

There may be a bigger question about how we intend users to input sensitive metadata - for now, the wizard doesn't support distinguishing between non-sensitive and sensitive. We may need to instruct them to go to the web portal to write in anything sensitive.

Reply:
Good point, though we do need the ability to create sensitive metadata through the wizard. I am inclined to leave it the way it is and warn our users. The number of projects with sensitive metadata is likely to be small to start with and I think we may be better able to deal with this when we migrate to server side